applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py

```diff
@@ -0,0 +1,185 @@
+from __future__ import print_function, absolute_import, division
+
+import time as timer
+
+import KratosMultiphysics
+from KratosMultiphysics import Logger
+
+from .response_function import ResponseFunctionBase
+
+import numpy as np
+
+class ValueAndGradient(object):
+
+    def __init__(self, infeasible_side):
+        self.infeasible_side = infeasible_side
+
+    def GetValue(self, distance, normal):
+        if self.infeasible_side and distance > 0:
+            return self._GetValue(distance, normal)
+        elif not self.infeasible_side and distance < 0:
+            return self._GetValue(distance, normal)
+        return 0.0
+
+    def GetGradient(self, distance, normal):
+        if self.infeasible_side and distance > 0:
+            return self._GetGradient(distance, normal)
+        elif not self.infeasible_side and distance < 0:
+            return self._GetGradient(distance, normal)
+        return [0.0, 0.0, 0.0]
+
+    def _GetValue(self, distance, normal):
+        raise NotImplementedError()
+
+    def _GetGradient(self, distance, normal):
+        raise NotImplementedError()
+
+class ScalarDistancePow(ValueAndGradient):
+
+    exponent = 1
+
+    def _GetValue(self, distance, normal):
+        return pow(distance, self.exponent)
+
+    def _GetGradient(self, distance, normal):
+        factor = self.exponent * pow(distance, self.exponent-1)
+        gradient = [
+            normal[0] * factor,
+            normal[1] * factor,
+            normal[2] * factor
+        ]
+        return gradient
+
+class ComponentDistance(ValueAndGradient):
+    '''Is the same as ScalarDistancePow with exponent 1'''
+
+    def _GetValue(self, distance, normal):
+        return distance
+
+    def _GetGradient(self, distance, normal):
+        gradient = [
+            normal[0],
+            normal[1],
+            normal[2]
+        ]
+        return gradient
+
+class Exponential(object):
+
+    def _GetValue(self, distance, normal):
+        return np.exp(distance)
+
+    def _GetGradient(self, distance, normal):
+        factor = np.exp(distance)
+        gradient = [
+            normal[0] * factor,
+            normal[1] * factor,
+            normal[2] * factor
+        ]
+        return gradient
+
+class PackagingResponseBase(ResponseFunctionBase):
```

Comment:
From a general perspective: Doing it through the definition of a single response function already implies, that all distances are summed and combined to a single compromise function. This excluded an explicit treatment of all constraints separately. But anyways Kratos cannot handle such a situation right now. But what about "constraint aggregation", like a KS-aggregation. This should be doable in theory, right?

Reply:
Yes this can be done at this stage. In the end, the current implementation is nothing else than a squared-sum aggregation of all violations.

Reply:
But as discussed i simplified this a bit and removed the unused parts.